src/main/scala/org/coroutines/AstCanonicalization.scala

```diff
@@ -122,6 +121,48 @@ trait AstCanonicalization[C <: Context] {
       (decls, q"$localvarname")
     case q"$selector[..$tpts](...$paramss)" if tpts.length > 0 || paramss.length > 0 =>
       // application
+
+      // We first need to see which parameters, if any, are by-name. Those that are
+      // should not be canonicalized.
+
+      // Maps to the parameter lists, saying whether or not they are by-name.
+      val byNameParams: immutable.Seq[immutable.Seq[Boolean]] = {
+        // Check that `selector` has a non-trivial symbol. If it doesn't, we assume
+        // that there were no by-name parameters.
+        if (selector.symbol != null && selector.symbol != NoSymbol) {
+          val paramLists = selector.symbol.asMethod.paramLists
+
+          // This value does not contain repeated parameters.
+          val noRepeatedParamsSeq = paramLists.map { currentParamList =>
+            currentParamList.map { param => param.asTerm.isByNameParam }
+          }
+
+          // Check that parameters were both passed in to and requested by the method.
+          if (paramss.length > 0 && noRepeatedParamsSeq.length > 0) {
+            // Initially, we assume that no parameters were repeated.
+            var paramsSeq = noRepeatedParamsSeq
+
+            /** If the number of passed-in parameters is greater than the number of
+             *  parameters needed by the method, then we assume that there are repeated
+             *  parameters.
+             *
+             *  This problem is fixed by appending the last value in
+             *  `noRepeatedParamsSeq` to `paramsSeq`. This is done until the number of
+             *  parameters are equal.
+             */
+            while (paramss(0).length > paramsSeq(0).length) {
+              val newHead = paramsSeq(0) :+ paramsSeq(0).last
+              val newTail = paramsSeq.tail
+              paramsSeq = newHead :: newTail
+            }
+            paramsSeq
+          } else {
+            noRepeatedParamsSeq
+          }
+        } else {
+          immutable.Seq.fill(1, paramss(0).length)(false)
```

Comment:
Should it not be:

```
for (params <- paramss) yield
  for (p <- params) yield false
```

Otherwise, with the current code you'd end up having only the first parameter list.